Content.Server/_Goobstation/Teleportation/Systems/PocketDimensionSystem.cs

```diff
@@ -0,0 +1,124 @@
+    [Dependency] private readonly IMapManager _mapMan = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+
+    private ISawmill _sawmill = default!;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<PocketDimensionComponent, ComponentRemove>(OnRemoved);
+        SubscribeLocalEvent<PocketDimensionComponent, GetVerbsEvent<AlternativeVerb>>(OnGetVerbs);
+
+        _sawmill = Logger.GetSawmill("pocket_dimension");
+    }
+
+    private void OnRemoved(EntityUid uid, PocketDimensionComponent comp, ComponentRemove args)
+    {
+        if (!Deleted(comp.PocketDimensionMap))
+            QueueDel(comp.PocketDimensionMap.Value);
+    }
+
+    private void OnGetVerbs(EntityUid uid, PocketDimensionComponent comp, GetVerbsEvent<AlternativeVerb> args)
+    {
+        if (!args.CanAccess || !args.CanInteract || !HasComp<HandsComponent>(args.User))
+            return;
+
+        AlternativeVerb verb = new()
+        {
+            Text = Loc.GetString("pocket-dimension-verb-text"),
+            Act = () => HandleActivation(uid, comp, args.User)
+        };
+        args.Verbs.Add(verb);
+    }
+
+    /// <summary>
+    /// Handles toggling the portal to the pocket dimension.
+    /// </summary>
+    private void HandleActivation(EntityUid uid, PocketDimensionComponent comp, EntityUid user)
+    {
+        if (Deleted(comp.PocketDimensionMap))
+        {
+            var map = _mapMan.CreateMap();
+
+            if (!_mapLoader.TryLoad(map, comp.PocketDimensionPath.ToString(), out var roots))
+            {
+                _sawmill.Error($"Failed to load pocket dimension map {comp.PocketDimensionPath}");
+                QueueDel(_mapMan.GetMapEntityId(map));
+                return;
+            }
+
+            comp.PocketDimensionMap = _mapMan.GetMapEntityId(map);
+
+            // find the pocket dimension's first grid and put the portal there
+            bool foundGrid = false;
+            foreach (var root in roots)
+            {
+                if (!HasComp<MapGridComponent>(root))
+                    continue;
+
+                // spawn the permanent portal into the pocket dimension, now ready to be used
+                var pos = Transform(root).Coordinates;
+                comp.ExitPortal = Spawn(comp.ExitPortalPrototype, pos);
+                EnsureComp<PortalComponent>(comp.ExitPortal!.Value, out var portal);
+                // the TryUnlink cleanup when first trying to create portal will fail without this
+                EnsureComp<LinkedEntityComponent>(uid);
+                portal.CanTeleportToOtherMaps = true;
+
+                _sawmill.Info($"Created pocket dimension on grid {root} of map {map}");
+
+                // if someone closes your portal you can use the one inside to escape
+                _link.OneWayLink(comp.ExitPortal.Value, uid);
+                foundGrid = true;
+            }
```

Comment:
```suggestion
                foundGrid = true;
                break;
            }
```
Maybe you want to break out of the loop if you found grid.